vortex-array/src/array/primitive/mod.rs

```diff
@@ -191,3 +194,184 @@ impl<'a> Array<'a> {
 }
 
 impl EncodingCompression for PrimitiveEncoding {}
+
+impl ScalarSubtractFn for PrimitiveArray<'_> {
+    fn scalar_subtract(&self, to_subtract: &Scalar) -> VortexResult<OwnedArray> {
+        if !self.dtype().eq_ignore_nullability(to_subtract.dtype()) {
+            vortex_bail!(MismatchedTypes: self.dtype(), to_subtract.dtype())
+        }
+
+        let should_wrap = match self.validity() {
+            Validity::AllInvalid => {
+                return Ok(self.clone().into_array().to_static());
+            }
+            Validity::NonNullable | Validity::AllValid => false,
+            Validity::Array(_) => true,
+        };
+
+        let result = match to_subtract.dtype() {
+            DType::Int(..) => {
+                match_each_integer_ptype!(self.ptype(), |$T| {
+                    let to_subtract = $T::try_from(to_subtract)?;
+                    let maybe_min = self.statistics().compute_as_cast(Stat::Min);
+
+                    if let Some(min) = maybe_min {
+                        let min: $T = min;
+                        if let (min, true) = min.overflowing_sub(to_subtract) {
+                            vortex_bail!("Integer subtraction over/underflow: {}, {}", min, to_subtract)
+                        }
+                        if let Some(max) = self.statistics().compute_as_cast(Stat::Max) {
+                            let max: $T = max;
+                            if let (max, true) = max.overflowing_sub(to_subtract) {
+                                vortex_bail!("Integer subtraction over/underflow: {}, {}", max, to_subtract)
+                            }
+                        }
+                    }
+
+                    let sub_vec : Vec<$T> = if should_wrap {
+                        self.typed_data::<$T>().iter().map(|&v| $T::checked_sub(v, to_subtract))
```

Comment:
Instead of this, you could `typed_data().iter().zip(logical_validity.to_null_buffer())` so you can handle validity explicitly. This way you can ensure all nulls are internally zeros, instead of just zero when they overflow.

The internal nulls aren't logically important, but they do impact compression ratios.